idx/bigtable/bigtable.go

```diff
@@ -0,0 +1,550 @@
+	statPruneDuration = stats.NewLatencyHistogram15s32("idx.bigtable.prune")
+	// metric idx.bigtable.delete is the duration of a delete of one or more metrics from the bigtable idx, including the delete from the in-memory index and the delete query
+	statDeleteDuration = stats.NewLatencyHistogram15s32("idx.bigtable.delete")
+	// metric idx.bigtable.save.skipped is how many saves have been skipped due to the writeQueue being full
+	statSaveSkipped = stats.NewCounter32("idx.bigtable.save.skipped")
+	// metric idx.bigtable.save.bytes-per-request is the number of bytes written to bigtable in each request.
+	statSaveBytesPerRequest = stats.NewMeter32("idx.bigtable.save.bytes-per-request", true)
+
+	Enabled           bool
+	maxStale          time.Duration
+	pruneInterval     time.Duration
+	updateBigTableIdx bool
+	updateInterval    time.Duration
+	updateInterval32  uint32
+
+	writeQueueSize    int
+	writeConcurrency  int
+	writeMaxFlushSize int
+
+	bigtableInstance string
+	gcpProject       string
+	tableName        string
+	createCF         bool
+)
+
+func ConfigSetup() {
+	btIdx := flag.NewFlagSet("bigtable-idx", flag.ExitOnError)
+
+	btIdx.BoolVar(&Enabled, "enabled", false, "")
+	btIdx.StringVar(&gcpProject, "gcp-project", "default", "Name of GCP project the bigtable cluster resides in")
+	btIdx.StringVar(&bigtableInstance, "bigtable-instance", "default", "Name of bigtable instance")
+	btIdx.StringVar(&tableName, "table-name", "metric_idx", "Name of bigtable table used for metricDefs")
+	btIdx.IntVar(&writeQueueSize, "write-queue-size", 100000, "Max number of metricDefs allowed to be unwritten to bigtable. Must be larger then write-max-flush-size")
+	btIdx.IntVar(&writeConcurrency, "write-concurrency", 5, "Number of writer threads to use")
+	btIdx.IntVar(&writeMaxFlushSize, "write-max-flush-size", 10000, "Max number of metricDefs in each batch write to bigtable")
+	btIdx.BoolVar(&updateBigTableIdx, "update-bigtable-index", true, "synchronize index changes to bigtable. not all your nodes need to do this.")
+	btIdx.DurationVar(&updateInterval, "update-interval", time.Hour*3, "frequency at which we should update the metricDef lastUpdate field, use 0s for instant updates")
+	btIdx.DurationVar(&maxStale, "max-stale", 0, "clear series from the index if they have not been seen for this much time.")
+	btIdx.DurationVar(&pruneInterval, "prune-interval", time.Hour*3, "Interval at which the index should be checked for stale series.")
+	btIdx.BoolVar(&createCF, "create-cf", true, "enable the creation of the table and column families")
+
+	globalconf.Register("bigtable-idx", btIdx)
+	return
+}
+
+func ConfigProcess() {
+	return
+}
+
+type writeReq struct {
+	def      *schema.MetricDefinition
+	recvTime time.Time
+}
+
+type BigtableIdx struct {
+	memory.MemoryIdx
+	tbl        *bigtable.Table
+	client     *bigtable.Client
+	writeQueue chan writeReq
+	shutdown   chan struct{}
+	wg         sync.WaitGroup
+}
+
+func New() *BigtableIdx {
```

Comment:
let's stay consistent with Cass Idx:
* New is for construction of the structure, not for initializing
* InitBare is for creating the session and assuring the schema's exist etc
* Init: inits the MemoryIdx, calls InitBare, creates the write workers and pruning routine, rebuilds the index, etc